src/stan-users-guide/regression.Rmd

```diff
@@ -1479,45 +1479,50 @@ factor of the final covariance matrix,
 ```
 Sigma_beta
   = quad_form_diag(Omega, tau)
-  = diag_post_multiply(L_Omega,tau) * diag_post_multiply(L_Omega,tau)'
+  = diag_pre_multiply(tau, L_Omega) * diag_pre_multiply(tau, L_Omega)'
 ```
 
-where the diagonal post-multiply compound operation is defined by
+where the diagonal pre-multiply compound operation is defined by
 
 ```
-diag_post_multiply(a, b) = a * diag_matrix(b)
+diag_pre_multiply(a, b) = diag_matrix(a) * b
 ```
 
 The new variable `z` is declared as a matrix, the entries of
 which are given independent standard normal priors; the `to_vector`
 operation turns the matrix into a vector so that it can be used as a
-vectorized argument to the univariate normal density.  Multiplying the
-Cholesky factor of the covariance matrix by `z` and adding the
-mean `(u * gamma)'` produces a `beta` distributed as in
-the original model.
+vectorized argument to the univariate normal density.  This results in every
+column of `z` being a $K$-variate normal random vector with the identity as
+covariance matrix.  Therefore, multiplying `z` by the Cholesky factor of the
+covariance matrix and adding the mean `(u * gamma)'` produces a `beta`
+distributed as in the original model, since
 
-Omitting the data declarations, which are the same as before, the
+$$
+\mathbb{V}[\beta] = \mathrm{diag}(\tau)\,\Omega_L\,\mathbb{V}[z]\,(\mathrm{diag}(\tau)\,\Omega_L)' = \mathrm{diag}(\tau)\,\Omega\,\mathrm{diag}(\tau) = \Sigma.
+$$
+
+Omitting the remaining data declarations, which are the same as before, the
 optimized model is as follows.
 
 ```
 parameters {
-  matrix[J, K] z;
+  matrix[K, J] z;
   cholesky_factor_corr[K] L_Omega;
   vector<lower=0,upper=pi()/2>[K] tau_unif;
-  matrix[L, K] gamma;                         // group coeffs
-  real<lower=0> sigma;                       // prediction error scale
+  matrix[L, K] gamma;                                         // group coeffs
+  real<lower=0> sigma;                                        // prediction error scale
 }
 transformed parameters {
-  matrix[J, K] beta;
-  vector<lower=0>[K] tau;     // prior scale
+  matrix[K, J] beta;
+  vector<lower=0>[K] tau;                                     // prior scale
   for (k in 1:K) tau[k] = 2.5 * tan(tau_unif[k]);
-  beta = u * gamma + z * diag_post_multiply(L_Omega,tau);
+  beta = u * gamma + diag_pre_multiply(tau, L_Omega) * z;
 }
 model {
   to_vector(z) ~ std_normal();
   L_Omega ~ lkj_corr_cholesky(2);
   to_vector(gamma) ~ normal(0, 5);
-  y ~ normal(rows_dot_product(beta[jj] , x), sigma);
+  y ~ normal(columns_dot_product(x', beta[, jj]), sigma);
```

Comment:
Woof this `columns_dot_product` is hard to read.

I see it was there originally, but any change the more explicit version is faster?:

```
vector[N] mu;

for(n in 1:N) {
  mu[n] = x[n, ] * beta[, jj[n]];
}

y ~ normal(mu, sigma);
```

Internally `columns_dot_product` is implemented as a bunch of dot products anyway. Duplicating `beta` out with `beta[, jj]` seems really inefficient and also hard to read.

Reply:
Thank you for reviewing the commit. I know too little about the respective implementation to answer that, but I agree that it is hard to read, which at least didactically is a downside. If you can point me in the direction of how to benchmark this kind of stuff easily, I would be willing to try.

In any case, from what you say I highly suspect the difference to be negligible anyway, so I think the explicit version is probably the better choice.

Reply:
> If you can point me in the direction of how to benchmark this kind of stuff easily

Did you have a model you were using this for? If you just run it and it looks faster to you, that's good enough for me. If nothing changes (and it's not obvious from just running the model a couple times), no biggie, we'll stick with what was there.

Reply:
Yes but not this part, only the one about the Cholesky parameterization.

To benchmark I now tried generating data and fitting the model, but something is wrong and I can't recover parameters and get bad Rhats and I don't understand what's wrong. I will open a Discourse thread for that and try to get to the bottom of this before I can go back to testing the speed of the two different implementations.

Reply:
Quick update on that. I did the benchmark for the two versions separately via
`
model_code <-
  "functions {
    real cdp_ver(int N, int[] jj, vector y, matrix x, matrix beta, real sigma) {
      return(normal_lpdf(y | columns_dot_product(x', beta[, jj]), sigma));
    }
    real exp_ver(int N, int[] jj, vector y, matrix x, matrix beta, real sigma) {
      vector[N] mu;
      
      for(n in 1:N) {
        mu[n] = x[n, ] * beta[, jj[n]];
      }
      
      return(normal_lpdf(y | mu, sigma));
    }
  }"

expose_stan_functions(stanc(model_code = model_code))

N <- 1000
K <- 10^5
J <- 100
jj <- sample(1:J, N, replace = TRUE)
y <- rnorm(N)
x <- matrix(rnorm(N * K), N, K)
beta <- matrix(rnorm(K * J), K, J)
sigma <- 1

lapply(
  `names<-`(10^c(1:5), sprintf("K = %i", 10^c(1:5))),
  function(this_K) {
    this_x <- x[, 1:this_K]
    this_beta <- beta[1:this_K, ]
    
    microbenchmark::microbenchmark(
      cdp_ver(N, jj, y, this_x, this_beta, sigma),
      exp_ver(N, jj, y, this_x, this_beta, sigma)
    )
  }
)
`

and these are the results:

`
$`K = 10`
Unit: microseconds
                                        expr    min     lq     mean median     uq     max neval
 cdp_ver(N, jj, y, this_x, this_beta, sigma) 31.980 32.642 34.21178 33.017 33.288 142.259   100
 exp_ver(N, jj, y, this_x, this_beta, sigma) 53.882 54.668 55.44646 55.159 55.590  71.154   100

$`K = 100`
Unit: microseconds
                                        expr     min       lq     mean  median      uq     max neval
 cdp_ver(N, jj, y, this_x, this_beta, sigma) 177.546 179.0475 187.5533 180.135 181.903 545.879   100
 exp_ver(N, jj, y, this_x, this_beta, sigma) 251.204 256.1485 259.0737 257.656 259.530 299.615   100

$`K = 1000`
Unit: milliseconds
                                        expr      min       lq     mean   median       uq      max neval
 cdp_ver(N, jj, y, this_x, this_beta, sigma) 3.374105 4.084625 4.303603 4.297131 4.518158 5.299637   100
 exp_ver(N, jj, y, this_x, this_beta, sigma) 2.736161 3.073163 3.185920 3.230519 3.316341 3.710950   100

$`K = 10000`
Unit: milliseconds
                                        expr      min       lq     mean   median       uq      max neval
 cdp_ver(N, jj, y, this_x, this_beta, sigma) 68.82494 71.22474 73.19782 72.67416 74.50988 81.12766   100
 exp_ver(N, jj, y, this_x, this_beta, sigma) 67.90834 70.13592 71.35018 71.25214 72.40257 78.54489   100

$`K = 100000`
Unit: milliseconds
                                        expr      min       lq     mean   median       uq      max neval
 cdp_ver(N, jj, y, this_x, this_beta, sigma) 599.8329 619.4768 624.4853 624.4542 629.9424 641.3602   100
 exp_ver(N, jj, y, this_x, this_beta, sigma) 441.7484 461.2230 472.5214 472.8654 484.1613 508.1761   100
`

So all else being equal, it seems to depend on the size of K (do you have an idea why this might be?). Since the cdp version is only faster for small K on the microsecond scale and the explicit version is faster for bigger K (so where it matters more I would say), maybe it would indeed be best to use the explicit version.

Reply:
Thanks for the benchmark.

> Since the cdp version is only faster for small K on the microsecond scale and the explicit version is faster for bigger K (so where it matters more I would say), maybe it would indeed be best to use the explicit version.

Sounds good to me.

I checked holding K fixed and changing N and similar thing happens.

```
N <- 100000
K <- 100
J <- 100
y <- rnorm(N)
x <- matrix(rnorm(N * K), N, K)
beta <- matrix(rnorm(K * J), K, J)
sigma <- 1

lapply(
  `names<-`(10^c(1:5), sprintf("N = %i", 10^c(1:5))),
  function(this_N) {
    this_x <- x[1:this_N, ]
    this_y <- y[1:this_N]
    jj <- sample(1:J, this_N, replace = TRUE)
    
    microbenchmark::microbenchmark(
      cdp_ver(this_N, jj, this_y, this_x, beta, sigma),
      exp_ver(this_N, jj, this_y, this_x, beta, sigma)
    )
  }
)
```

```
$`N = 10`
Unit: microseconds
                                             expr   min    lq    mean median    uq    max neval
 cdp_ver(this_N, jj, this_y, this_x, beta, sigma) 6.231 6.412 6.80094  6.502 6.683 27.712   100
 exp_ver(this_N, jj, this_y, this_x, beta, sigma) 6.723 6.843 7.06449  6.948 7.093  9.939   100

$`N = 100`
Unit: microseconds
                                             expr    min     lq     mean  median    uq    max neval
 cdp_ver(this_N, jj, this_y, this_x, beta, sigma) 33.152 33.673 34.87254 34.3445 34.88 77.305   100
 exp_ver(this_N, jj, this_y, this_x, beta, sigma) 40.817 41.708 42.34889 42.1090 42.62 46.688   100

$`N = 1000`
Unit: microseconds
                                             expr     min      lq     mean   median       uq     max neval
 cdp_ver(this_N, jj, this_y, this_x, beta, sigma) 228.379 293.856 290.3430 295.9095 299.4210 651.782   100
 exp_ver(this_N, jj, this_y, this_x, beta, sigma) 302.998 314.339 370.5593 387.0610 390.7475 528.040   100

$`N = 10000`
Unit: milliseconds
                                             expr      min       lq     mean   median       uq       max neval
 cdp_ver(this_N, jj, this_y, this_x, beta, sigma) 7.697772 7.778669 8.581241 9.135988 9.203350 10.093995   100
 exp_ver(this_N, jj, this_y, this_x, beta, sigma) 3.869074 3.929543 4.389059 4.257292 4.814692  5.150016   100

$`N = 100000`
Unit: milliseconds
                                             expr      min       lq     mean   median       uq      max neval
 cdp_ver(this_N, jj, this_y, this_x, beta, sigma) 84.27250 84.82930 84.95171 84.91227 85.02865 88.49719   100
 exp_ver(this_N, jj, this_y, this_x, beta, sigma) 52.83894 53.35463 53.88684 53.43763 53.49507 64.63036   100
```